math/src/circle/polynomial.rs

```diff
@@ -0,0 +1,287 @@
+extern crate alloc;
+#[cfg(feature = "alloc")]
+use super::{
+    cfft::{cfft, icfft, order_cfft_result_naive, order_icfft_input_naive},
+    cosets::Coset,
+    twiddles::{get_twiddles, TwiddlesConfig},
+};
+use crate::{
+    fft::cpu::bit_reversing::in_place_bit_reverse_permute,
+    field::{element::FieldElement, fields::mersenne31::field::Mersenne31Field},
+};
+use alloc::vec::Vec;
+
+/// Given the 2^n coefficients of a two-variables polynomial of degree 2^n - 1 in the basis {1, y, x, xy, 2xˆ2 -1, 2xˆ2y-y, 2xˆ3-x, 2xˆ3y-xy,...}
+/// returns the evaluation of the polynomial on the points of the standard coset of size 2^n.
+/// Note that coeff has to be a vector with length a power of two 2^n.
+#[cfg(feature = "alloc")]
+pub fn evaluate_cfft(
+    mut coeff: Vec<FieldElement<Mersenne31Field>>,
+) -> Vec<FieldElement<Mersenne31Field>> {
+    // We get the twiddles for the Evaluation.
+    let domain_log_2_size: u32 = coeff.len().trailing_zeros();
+    let coset = Coset::new_standard(domain_log_2_size);
+    let config = TwiddlesConfig::Evaluation;
+    let twiddles = get_twiddles(coset, config);
+
+    // For our algorithm to work, we must give as input the coefficients in bit reverse order.
+    in_place_bit_reverse_permute::<FieldElement<Mersenne31Field>>(&mut coeff);
+    cfft(&mut coeff, twiddles);
+
+    // The cfft returns the evaluations in a certain order, so we permute them to get the natural order.
+    order_cfft_result_naive(&mut coeff)
+}
+
+/// Interpolates the 2^n evaluations of a two-variables polynomial of degree 2^n - 1 on the points of the standard coset of size 2^n.
+/// As a result we obtain the coefficients of the polynomial in the basis: {1, y, x, xy, 2xˆ2 -1, 2xˆ2y-y, 2xˆ3-x, 2xˆ3y-xy,...}
+/// Note that eval has to be a vector of length a power of two 2^n.
+#[cfg(feature = "alloc")]
+pub fn interpolate_cfft(
```

Comment:
As an optimization, we could just send the twiddles as input to the function. This is useful when we have to interpolate multiple columns